pkg/image/trigger/annotations/annotations.go

```diff
@@ -0,0 +1,266 @@
+	spec, path, err := ometa.GetPodSpec(obj)
+	if err != nil {
+		return nil, ""
+	}
+	relative := strings.TrimPrefix(fieldPath, path.String()+".")
+	if relative == fieldPath {
+		return nil, ""
+	}
+	return findPodSpecContainer(spec, relative)
+}
+
+// UpdateObjectFromImages attempts to set the appropriate object information. If changes are necessary, it lazily copies
+// obj and returns it, or if no changes are necessary returns nil.
+func UpdateObjectFromImages(obj runtime.Object, copier runtime.ObjectCopier, tagRetriever trigger.TagRetriever) (runtime.Object, error) {
+	var updated runtime.Object
+	m, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, err
+	}
+	spec, path, err := ometa.GetPodSpec(obj)
+	if err != nil {
+		return nil, err
+	}
+	basePath := path.String() + "."
+	_, _, triggers, err := calculateAnnotationTriggers(m, "/")
+	if err != nil {
+		return nil, err
+	}
+	for _, trigger := range triggers {
+		fieldPath := trigger.FieldPath
+		if !strings.HasPrefix(trigger.FieldPath, basePath) {
+			continue
+		}
+		fieldPath = strings.TrimPrefix(fieldPath, basePath)
+
+		namespace := trigger.From.Namespace
+		if len(namespace) == 0 {
+			namespace = m.GetNamespace()
+		}
+		ref, _, ok := tagRetriever.ImageStreamTag(namespace, trigger.From.Name)
+		if !ok {
+			continue
+		}
+
+		container, _ := findPodSpecContainer(spec, fieldPath)
+		if container.Image != ref {
+			if updated == nil {
+				updated, err = copier.Copy(obj)
+				if err != nil {
+					return nil, err
+				}
+				spec, _, _ = ometa.GetPodSpec(updated)
+				container, _ = findPodSpecContainer(spec, fieldPath)
+			}
+			container.Image = ref
+		}
+	}
+	return updated, nil
+}
+
+// annotationTriggerIndexer uses annotations on objects to trigger changes.
+type annotationTriggerIndexer struct {
+	prefix string
+}
+
+// NewAnnotationTriggerIndexer creates an indexer that deals with objects that have a pod spec and use
+// annotations to indicate the desire to trigger.
+func NewAnnotationTriggerIndexer(prefix string) trigger.Indexer {
+	return annotationTriggerIndexer{prefix: prefix}
+}
+
+func (i annotationTriggerIndexer) Index(obj, old interface{}) (string, *trigger.CacheEntry, cache.DeltaType, error) {
+	var (
+		triggers  []triggerapi.ObjectFieldTrigger
+		key       string
+		namespace string
+		change    cache.DeltaType
+	)
+	switch {
+	case obj != nil && old == nil:
+		// added
+		m, err := meta.Accessor(obj)
+		if err != nil {
+			return "", nil, change, err
+		}
+		key, namespace, triggers, err = calculateAnnotationTriggers(m, i.prefix)
+		if err != nil {
+			return "", nil, change, err
+		}
+		change = cache.Added
+	case old != nil && obj == nil:
+		// deleted
+		m, err := meta.Accessor(old)
+		if err != nil {
+			return "", nil, change, err
+		}
+		key, namespace, triggers, err = calculateAnnotationTriggers(m, i.prefix)
+		if err != nil {
+			return "", nil, change, err
+		}
+		change = cache.Deleted
+	default:
```

Comment:
When does informer return two nil objects?  Isn't it required to return the cache stub?